### PR TITLE
Support for Ghost Pointers

### PR DESCRIPTION
**Pull request by ArquintL**
This PR adds support for ghost pointers. However, ghost fields will be added by a separate PR. Note that this PR does not address the issue of calling ghost functions in non-ghost code, which is instead part of PR #755.

### CODE REVIEW

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala [hunk 1]
```diff
@@ -689,7 +689,7 @@ trait ExprTyping extends BaseTyping { this: TypeInfoImpl =>
       (underlyingType(baseType), underlyingType(idxType)) match {
         case (Single(base), Single(idx)) => (base, idx) match {
           case (ArrayT(_, elem), IntT(_)) => elem
-          case (PointerT(ArrayT(_, elem)), IntT(_)) => elem
+          case (ActualPointerT(ArrayT(_, elem)), IntT(_)) => elem
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala:692]: (repeat) I think a ghost pointer to an array is ok, i.e. this line should account for ghost pointer too.

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala [hunk 2]
```diff
@@ -390,7 +390,7 @@ trait ExprTyping extends BaseTyping { this: TypeInfoImpl =>
             error(cap, "sequence slice expressions do not allow specifying a capacity", capT.isDefined)
         }
 
-        case (PointerT(ArrayT(l, _)), None | Some(IntT(_)), None | Some(IntT(_)), None | Some(IntT(_))) =>
+        case (ActualPointerT(ArrayT(l, _)), None | Some(IntT(_)), None | Some(IntT(_)), None | Some(IntT(_))) =>
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala:393]:
```suggestion
        case (ActualPointerT(ArrayT(l, _)), None | Some(IntT(_)), None | Some(IntT(_)), None | Some(IntT(_))) =>  // without ghost slices, slicing a ghost pointer is not allowed.
```

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala [hunk 3]
```diff
@@ -324,7 +324,7 @@ trait ExprTyping extends BaseTyping { this: TypeInfoImpl =>
               val idxOpt = intConstantEval(index)
               error(n, s"index $index is out of bounds", !idxOpt.forall(i => i >= 0 && i < l))
 
-            case (PointerT(ArrayT(l, _)), IntT(_)) =>
+            case (ActualPointerT(ArrayT(l, _)), IntT(_)) =>
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala:327]: I think a ghost pointer to an array is ok, i.e. this line should account for ghost pointer too.

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/MiscTyping.scala [hunk 4]
```diff
@@ -25,8 +25,8 @@ trait MiscTyping extends BaseTyping { this: TypeInfoImpl =>
   private[typing] def wellDefActualMisc(misc: PActualMisc): Messages = misc match {
 
     case n@PRange(exp, _) => isExpr(exp).out ++ (underlyingType(exprType(exp)) match {
-      case _: ArrayT | PointerT(_: ArrayT) | _: SliceT | _: GhostSliceT | _: MapT |
-          ChannelT(_, ChannelModus.Recv | ChannelModus.Bi) => noMessages
+      case _: ArrayT | ActualPointerT(_: ArrayT) | _: SliceT | _: GhostSliceT | _: MapT |
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/typing/MiscTyping.scala:28]: ranging over an array ghost pointer is ok in ghost code.

File: src/test/resources/regressions/features/ghost_pointer/pointer-creation-fail01.gobra [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    // gpointer is a ghost type which cannot be used for actual variables:
+    //:: ExpectedOutput(type_error)
+    var y gpointer[int]
+
+    ghost var z gpointer[int]
+
+    // ghost pointer is not assignable to an actual pointer
+    //:: ExpectedOutput(type_error)
+    ghost x = new(int)
```
Comment (Felalolf) [inline src/test/resources/regressions/features/ghost_pointer/pointer-creation-fail01.gobra:29]:
```suggestion
    ghost x = new(int)
    
    ghost z = new(int)
```

**Review by Felalolf — APPROVED** [review]
Looks good, I only have some minor comments. Please adapt the index expression to also work with array ghost pointers before you merge.

**Comment by ArquintL** [comment]
I had to change the implementation quite a bit more and introduce the concept of whether the location an expression refers to is ghost or not. Please take another look

File: src/test/resources/regressions/issues/000039.gobra [hunk 6]
```diff
@@ -8,7 +8,8 @@ func test() {
   x1  := 5
   x2@ := 5
 
-  ghost L:
+  // an empty block is inserted such that the assignments are not treated being part of the ghost labeled statement
+  ghost L: {}
```
Comment (Felalolf) [inline src/test/resources/regressions/issues/000039.gobra:12]: Oh, wow. You do not need to adapt anything, but good to be aware of this.

File: src/test/resources/regressions/features/ghost_pointer/ghost-reference-simple01.gobra [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+type Test struct {
+    field int
+}
+
+ghost
+decreases
+ensures acc(x)
+func GhostFunc() (x gpointer[Test]) {
+    x = &Test{ 42 }
```
Comment (Felalolf) [inline src/test/resources/regressions/features/ghost_pointer/ghost-reference-simple01.gobra:14]: Do we support `x = nil`? If yes, then add a test for that. If no, then we might want to add this later (can be done in a later PR).

File: src/main/scala/viper/gobra/frontend/info/implementation/resolution/MemberResolution.scala [hunk 8]
```diff
@@ -86,7 +86,7 @@ trait MemberResolution { this: TypeInfoImpl =>
     def go(pastDeref: Boolean): Type => AdvancedMemberSet[AdtMember] = attr[Type, AdvancedMemberSet[AdtMember]] {
 
       case DeclaredT(decl, context) => go(pastDeref)(context.symbType(decl.right)).surface
-      case PointerT(t) if !pastDeref => go(pastDeref = true)(t).ref
+      case ActualPointerT(t) if !pastDeref => go(pastDeref = true)(t).ref
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/resolution/MemberResolution.scala:89]: If we have ghost pointer receiver, then you should duplicate the cases for ActualPointer also for GhostPointer (in this entire file). Doing the duplication should implement the expected member propagation for ghost pointer.

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostAssignability.scala [hunk 9]
```diff
@@ -103,17 +106,20 @@ trait GhostAssignability {
...
       case _ => error(left, "ghost error: selections on types are not assignable")
-  }
+    }
 
     case PBlankIdentifier() => noMessages
   }
 
+  // handles the case of assignments in ghost code
+  private def ghostAssigneeAssignmentMsgInGhostCode(left: PAssignee): Messages =
+    error(left, s"ghost error: only ghost locations can be assigned to in ghost code}", !ghostLocationClassification(left))
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostAssignability.scala:121]:
```suggestion
    error(left, s"ghost error: only ghost locations can be assigned to in ghost code", !ghostLocationClassification(left))
```

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostTyping.scala [hunk 10]
```diff
@@ -119,6 +120,44 @@ trait GhostTyping extends GhostClassifier { this: TypeInfoImpl =>
...
+    createGhostClassification[PExpression](e => ghostLocationTyping(e).isGhost)
+
+  /** returns ghost typing of the location that expression represents */
+  private[separation] lazy val ghostLocationTyping: PExpression => GhostType = {
+    import GhostType._
+
+    createGhostTyping[PExpression] {
+      case PNamedOperand(id) => ghost(ghostIdClassification(id))
+      case e: PDeref => resolve(e) match {
+        case Some(ap.Deref(base)) => exprType(base) match {
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostTyping.scala:134]: underlying type is missing

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostTyping.scala [hunk 11]
```diff
@@ -119,6 +120,44 @@ trait GhostTyping extends GhostClassifier { this: TypeInfoImpl =>
...
+          case _: Type.GhostPointerT => isGhost
+          case _ => notGhost
+        }
+        case _ => Violation.violation(s"type-checker should only permit PDeref assignees that have the Deref pattern")
+      }
+      case e: PDot => resolve(e) match {
+        case Some(s: ap.FieldSelection) =>
+          val isGhostField = ghostIdClassification(s.id)
+          (typ(s.base), isGhostField) match {
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostTyping.scala:143]: underlying type is missing

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostTyping.scala [hunk 12]
```diff
@@ -119,6 +120,44 @@ trait GhostTyping extends GhostClassifier { this: TypeInfoImpl =>
...
+            case (_, true) => isGhost // ghost fields are always ghost memory
+            case (tb: Type.PointerT, _) => ghost(tb.isInstanceOf[Type.GhostPointerT]) // (implicitly) dereferencing a ghost pointer leads to a ghost heap location
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostTyping.scala:145]:
```suggestion
            case (tb: Type.PointerT, _) => ghost(tb.isInstanceOf[Type.GhostPointerT]) // (implicitly) dereferencing a field of a ghost pointer leads to a ghost heap location
```

File: src/test/resources/regressions/issues/000039.gobra (hunk 6 above)
Comment (ArquintL) [reply src/test/resources/regressions/issues/000039.gobra:12]: It's unclear how we even would fix this as it seems to mostly be a problem in our head. Our mental model of a Viper label clashes with Go's labeled statement, which however only becomes an issue for ghost labels

File: src/test/resources/regressions/features/ghost_pointer/ghost-reference-simple01.gobra (hunk 7 above)
Comment (ArquintL) [reply src/test/resources/regressions/features/ghost_pointer/ghost-reference-simple01.gobra:14]: yes we do. I'm adding a testcase

**Pull request merged**